cpp/src/detail/shuffle_vertex_pairs.cu

```diff
@@ -160,96 +197,159 @@ shuffle_vertex_pairs_by_gpu_id_impl(
-      std::get<1>(*edge_id_type_tuple).resize(0, handle.get_stream());
-      std::get<1>(*edge_id_type_tuple).shrink_to_fit(handle.get_stream());
+    if (edge_types) {
+      std::tie(rx_edge_types, std::ignore) =
+        shuffle_values(comm, (*edge_types).begin(), h_tx_value_counts, handle.get_stream());
+      (*edge_types).resize(0, handle.get_stream());
+      (*edge_types).shrink_to_fit(handle.get_stream());
     }
   } else {
     if (weights) {
-      if (edge_id_type_tuple) {
-        std::forward_as_tuple(
-          std::tie(rx_majors, rx_minors, rx_weights, rx_edge_ids, rx_edge_types), std::ignore) =
-          shuffle_values(comm,
-                         thrust::make_zip_iterator(majors.begin(),
-                                                   minors.begin(),
-                                                   weights->begin(),
-                                                   std::get<0>(*edge_id_type_tuple).begin(),
-                                                   std::get<1>(*edge_id_type_tuple).begin()),
-                         h_tx_value_counts,
-                         handle.get_stream());
-        majors.resize(0, handle.get_stream());
-        majors.shrink_to_fit(handle.get_stream());
-        minors.resize(0, handle.get_stream());
-        minors.shrink_to_fit(handle.get_stream());
-        (*weights).resize(0, handle.get_stream());
-        (*weights).shrink_to_fit(handle.get_stream());
-        std::get<0>(*edge_id_type_tuple).resize(0, handle.get_stream());
-        std::get<0>(*edge_id_type_tuple).shrink_to_fit(handle.get_stream());
-        std::get<1>(*edge_id_type_tuple).resize(0, handle.get_stream());
-        std::get<1>(*edge_id_type_tuple).resize(0, handle.get_stream());
+      if (edge_ids) {
+        if (edge_types) {
+          std::forward_as_tuple(
+            std::tie(rx_majors, rx_minors, rx_weights, rx_edge_ids, rx_edge_types), std::ignore) =
+            shuffle_values(comm,
+                           thrust::make_zip_iterator(majors.begin(),
+                                                     minors.begin(),
+                                                     weights->begin(),
+                                                     edge_ids->begin(),
+                                                     edge_types->begin()),
+                           h_tx_value_counts,
+                           handle.get_stream());
+          majors.resize(0, handle.get_stream());
+          majors.shrink_to_fit(handle.get_stream());
+          minors.resize(0, handle.get_stream());
+          minors.shrink_to_fit(handle.get_stream());
+          (*weights).resize(0, handle.get_stream());
+          (*weights).shrink_to_fit(handle.get_stream());
+          (*edge_ids).resize(0, handle.get_stream());
+          (*edge_ids).shrink_to_fit(handle.get_stream());
+          (*edge_types).resize(0, handle.get_stream());
+          (*edge_types).shrink_to_fit(handle.get_stream());
```

Comment:
I think we can move this logic to the end of this function (right before return) to avoid repetition.

```
          majors.resize(0, handle.get_stream());
          majors.shrink_to_fit(handle.get_stream());
          minors.resize(0, handle.get_stream());
          minors.shrink_to_fit(handle.get_stream());
          if (weights) { ... }
          if (edge_ids) { ... }
          if (edge_types) { ... }
```

Reply:
Is there a reason to prefer this strategy (storing results in `rx_majors`, shrinking `majors` and later doing a `std::move` over the top) rather than just overwriting `majors` in the first place?

That is, I could do:
`std::tie(majors, minors, *weights, *edge_ids, *edge_types)` as the return for `shuffle_values` and it will just do an `std::move` over the top of those values.

Reply:
> Is there a reason to prefer this strategy

No, and I think what you described is better.